### PR TITLE
Fix bug #406: Get api_endpoints directly from the controller

### PR DESCRIPTION
**Pull request by davigar15**
This should fix the bug #406.

I also added a new command for getting ControllerAPIInformation.

### CODE REVIEW

File: juju/controller.py [hunk 1]
```diff
@@ -273,6 +274,16 @@ async def add_cloud(self, name, cloud):
...
 
+    async def info(self):
+        """Show Controller Info from connection
+
+        :return ControllerAPIInfoResult
+        """
+        log.debug('Getting information')
+        cloud_facade = client.CrossControllerFacade.from_connection(self.connection())
+        info = await cloud_facade.ControllerInfo()
+        return info
```
Comment (SimonRichardson) [inline juju/controller.py:285]: You can just return the await here.

Also can we get a test around the info?

Comment (davigar15) [reply juju/controller.py:285]: Yes!

Although I'm getting this error when testing it in OSM:
```
Exception: No facade CrossController in facades {'Action': 2, 'ActionPruner': 1, 'Agent': 2, 'AgentTools': 1, 'AllWatcher': 1, 'Annotations': 2, 'Application': 8, 'ApplicationScaler': 1, 'Backups': 2, 'Block': 2, 'Bundle': 3, 'CAASAgent': 1, 'CAASFirewaller': 1, 'CAASOperator': 1, 'CAASOperatorProvisioner': 1, 'CAASOperatorUpgrader': 1, 'CAASUnitProvisioner': 1, 'CharmRevisionUpdater': 2, 'Charms': 2, 'Cleaner': 2, 'Client': 2, 'CredentialManager': 1, 'CredentialValidator': 1, 'CrossModelRelations': 1, 'Deployer': 1, 'DiskManager': 2, 'EntityWatcher': 2, 'ExternalControllerUpdater': 1, 'FanConfigurer': 1, 'FilesystemAttachmentsWatcher': 2, 'FirewallRules': 1, 'Firewaller': 5, 'HighAvailability': 2, 'HostKeyReporter': 1, 'ImageManager': 2, 'ImageMetadata': 3, 'InstanceMutater': 2, 'InstancePoller': 3, 'KeyManager': 1, 'KeyUpdater': 1, 'LeadershipService': 2, 'LifeFlag': 1, 'LogForwarding': 1, 'Logger': 1, 'MachineActions': 1, 'MachineManager': 4, 'MachineUndertaker': 1, 'Machiner': 1, 'MeterStatus': 1, 'MetricsAdder': 2, 'MetricsDebug': 2, 'MetricsManager': 1, 'MigrationFlag': 1, 'MigrationMaster': 1, 'MigrationMinion': 1, 'MigrationStatusWatcher': 1, 'ModelConfig': 2, 'ModelGeneration': 2, 'ModelUpgrader': 1, 'NotifyWatcher': 1, 'OfferStatusWatcher': 1, 'Payloads': 1, 'PayloadsHookContext': 1, 'Pinger': 1, 'Provisioner': 6, 'ProxyUpdater': 2, 'Reboot': 2, 'RelationStatusWatcher': 1, 'RelationUnitsWatcher': 1, 'RemoteRelationWatcher': 1, 'RemoteRelations': 1, 'Resources': 1, 'ResourcesHookContext': 1, 'Resumer': 2, 'RetryStrategy': 1, 'SSHClient': 2, 'Singular': 2, 'Spaces': 3, 'StatusHistory': 2, 'Storage': 4, 'StorageProvisioner': 4, 'StringsWatcher': 1, 'Subnets': 2, 'Undertaker': 1, 'UnitAssigner': 1, 'Uniter': 8, 'UpgradeSeries': 1, 'UpgradeSteps': 1, 'Upgrader': 1, 'VolumeAttachmentPlansWatcher': 1, 'VolumeAttachmentsWatcher': 2}
2020-04-29T08:19:04 ERROR lcm.ns n2vc_juju_conn.py:251 Error creating machine on juju: Cannot get model 044736b4-ea83-4767-91af-db5cdc9065bf. Exception: No facade CrossController in facades {'Action': 2, 'ActionPruner': 1, 'Agent': 2, 'AgentTools': 1, 'AllWatcher': 1, 'Annotations': 2, 'Application': 8, 'ApplicationScaler': 1, 'Backups': 2, 'Block': 2, 'Bundle': 3, 'CAASAgent': 1, 'CAASFirewaller': 1, 'CAASOperator': 1, 'CAASOperatorProvisioner': 1, 'CAASOperatorUpgrader': 1, 'CAASUnitProvisioner': 1, 'CharmRevisionUpdater': 2, 'Charms': 2, 'Cleaner': 2, 'Client': 2, 'CredentialManager': 1, 'CredentialValidator': 1, 'CrossModelRelations': 1, 'Deployer': 1, 'DiskManager': 2, 'EntityWatcher': 2, 'ExternalControllerUpdater': 1, 'FanConfigurer': 1, 'FilesystemAttachmentsWatcher': 2, 'FirewallRules': 1, 'Firewaller': 5, 'HighAvailability': 2, 'HostKeyReporter': 1, 'ImageManager': 2, 'ImageMetadata': 3, 'InstanceMutater': 2, 'InstancePoller': 3, 'KeyManager': 1, 'KeyUpdater': 1, 'LeadershipService': 2, 'LifeFlag': 1, 'LogForwarding': 1, 'Logger': 1, 'MachineActions': 1, 'MachineManager': 4, 'MachineUndertaker': 1, 'Machiner': 1, 'MeterStatus': 1, 'MetricsAdder': 2, 'MetricsDebug': 2, 'MetricsManager': 1, 'MigrationFlag': 1, 'MigrationMaster': 1, 'MigrationMinion': 1, 'MigrationStatusWatcher': 1, 'ModelConfig': 2, 'ModelGeneration': 2, 'ModelUpgrader': 1, 'NotifyWatcher': 1, 'OfferStatusWatcher': 1, 'Payloads': 1, 'PayloadsHookContext': 1, 'Pinger': 1, 'Provisioner': 6, 'ProxyUpdater': 2, 'Reboot': 2, 'RelationStatusWatcher': 1, 'RelationUnitsWatcher': 1, 'RemoteRelationWatcher': 1, 'RemoteRelations': 1, 'Resources': 1, 'ResourcesHookContext': 1, 'Resumer': 2, 'RetryStrategy': 1, 'SSHClient': 2, 'Singular': 2, 'Spaces': 3, 'StatusHistory': 2, 'Storage': 4, 'StorageProvisioner': 4, 'StringsWatcher': 1, 'Subnets': 2, 'Undertaker': 1, 'UnitAssigner': 1, 'Uniter': 8, 'UpgradeSeries': 1, 'UpgradeSteps': 1, 'Upgrader': 1, 'VolumeAttachmentPlansWatcher': 1, 'VolumeAttachmentsWatcher': 2}
2020-04-29T08:19:04 ERROR lcm.ns ns.py:3000 Task ns=044736b4-ea83-4767-91af-db5cdc9065bf instantiate=3b6650ca-5874-4643-92fe-f84ce1fc7375 Deploying VCA 1.: create execution environment Error creating machine on juju: Cannot get model 044736b4-ea83-4767-91af-db5cdc9065bf. Exception: No facade CrossController in facades {'Action': 2, 'ActionPruner': 1, 'Agent': 2, 'AgentTools': 1, 'AllWatcher': 1, 'Annotations': 2, 'Application': 8, 'ApplicationScaler': 1, 'Backups': 2, 'Block': 2, 'Bundle': 3, 'CAASAgent': 1, 'CAASFirewaller': 1, 'CAASOperator': 1, 'CAASOperatorProvisioner': 1, 'CAASOperatorUpgrader': 1, 'CAASUnitProvisioner': 1, 'CharmRevisionUpdater': 2, 'Charms': 2, 'Cleaner': 2, 'Client': 2, 'CredentialManager': 1, 'CredentialValidator': 1, 'CrossModelRelations': 1, 'Deployer': 1, 'DiskManager': 2, 'EntityWatcher': 2, 'ExternalControllerUpdater': 1, 'FanConfigurer': 1, 'FilesystemAttachmentsWatcher': 2, 'FirewallRules': 1, 'Firewaller': 5, 'HighAvailability': 2, 'HostKeyReporter': 1, 'ImageManager': 2, 'ImageMetadata': 3, 'InstanceMutater': 2, 'InstancePoller': 3, 'KeyManager': 1, 'KeyUpdater': 1, 'LeadershipService': 2, 'LifeFlag': 1, 'LogForwarding': 1, 'Logger': 1, 'MachineActions': 1, 'MachineManager': 4, 'MachineUndertaker': 1, 'Machiner': 1, 'MeterStatus': 1, 'MetricsAdder': 2, 'MetricsDebug': 2, 'MetricsManager': 1, 'MigrationFlag': 1, 'MigrationMaster': 1, 'MigrationMinion': 1, 'MigrationStatusWatcher': 1, 'ModelConfig': 2, 'ModelGeneration': 2, 'ModelUpgrader': 1, 'NotifyWatcher': 1, 'OfferStatusWatcher': 1, 'Payloads': 1, 'PayloadsHookContext': 1, 'Pinger': 1, 'Provisioner': 6, 'ProxyUpdater': 2, 'Reboot': 2, 'RelationStatusWatcher': 1, 'RelationUnitsWatcher': 1, 'RemoteRelationWatcher': 1, 'RemoteRelations': 1, 'Resources': 1, 'ResourcesHookContext': 1, 'Resumer': 2, 'RetryStrategy': 1, 'SSHClient': 2, 'Singular': 2, 'Spaces': 3, 'StatusHistory': 2, 'Storage': 4, 'StorageProvisioner': 4, 'StringsWatcher': 1, 'Subnets': 2, 'Undertaker': 1, 'UnitAssigner': 1, 'Uniter': 8, 'UpgradeSeries': 1, 'UpgradeSteps': 1, 'Upgrader': 1, 'VolumeAttachmentPlansWatcher': 1, 'VolumeAttachmentsWatcher': 2}

```

Comment (davigar15) [reply juju/controller.py:285]: @SimonRichardson Any ideas?

File: juju/controller.py [hunk 2]
```diff
@@ -144,6 +149,11 @@ async def connect(self, *args, **kwargs):
                 raise ValueError('Authentication parameters are required '
                                  'if controller_name not given')
             await self._connector.connect(**kwargs)
+            info = await self.info()
+            self._connector._connection.endpoints =[
+                (e, info.results[0].cacert)
+                for e in info.results[0].addresses
+            ]
```
Comment (SimonRichardson) [inline juju/controller.py:156]: I'm going to be really annoying and say, can we move this after the `if...else...`. That way we only need to have this logic once.

File: test.py [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+        # import pdb; pdb.set_trace()
+        await asyncio.sleep(20)
+    await controller.disconnect()
+
+
+if __name__ == '__main__':
+    logging.basicConfig(level=logging.DEBUG)
+    ws_logger = logging.getLogger('websockets.protocol')
+    ws_logger.setLevel(logging.ERROR)
+    loop.run(main())
```
Comment (SimonRichardson) [inline test.py:77]: Delete this test.

File: juju/controller.py (hunk 2 above)
Comment (davigar15) [reply juju/controller.py:156]: sure! Totally makes sense!

File: test.py (hunk 3 above)
Comment (davigar15) [reply test.py:77]: ups

**Comment by SimonRichardson** [comment]
!!build!!

**Comment by SimonRichardson** [comment]
!!build!!

**Comment by SimonRichardson** [comment]
!!build!!

**Comment by SimonRichardson** [comment]
$$merge$$

**Pull request merged**